subprojects/core/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -117,4 +119,30 @@ public int hashCode() {
     public String toString() {
         return name;
     }
+
+    private static <T> void validateType(String name, Class<T> clazz) {
+        if (clazz.isArray()) {
+            validateType(name, clazz.getComponentType());
+            return;
+        }
+        if (Scalars.isScalarType(clazz)) {
+            return;
+        }
+        if (Named.class.isAssignableFrom(clazz)) {
+            return;
+        }
+        throw new IllegalArgumentException("Cannot declare a attribute '" + name + "' with type " + clazz + ". Supported types are: \n" + supportedTypes());
+    }
+
+    private static String supportedTypes() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("   - primitive types (byte, boolean, char, short, int, long, float, double) and their wrapped types (Byte, ...)\n");
```

Comment:
⭕️ Part of this concatenation could be moved to the `Scalars` class, using some kind of `describeTo(StringBuilder)` method.